client-misk-dynamodb/src/main/kotlin/app/cash/backfila/client/misk/dynamodb/internal/keyranges.kt

```diff
@@ -11,40 +20,148 @@ import okio.ByteString
 internal data class DynamoDbKeyRange(
   val start: Int,
   val end: Int,
-  val count: Int
-)
-
-internal fun decodeKeyRange(keyRange: KeyRange): DynamoDbKeyRange {
-  val (startOffset, count) = decodeSegment(keyRange.start)
-  val (endOffset, _) = decodeSegment(keyRange.end)
-  return DynamoDbKeyRange(startOffset, endOffset, count)
+  val count: Int,
+  val lastEvaluatedKey: Map<String, AttributeValue>? = null
+) {
+  init {
+    // Only allow lastEvaluatedKey to be non-null if the range includes a single segment.
+    if (lastEvaluatedKey != null) {
+      require(end == start + 1)
+    }
+  }
 }
 
-internal fun decodeSegment(segment: ByteString): Pair<Int, Int> {
-  val buffer = Buffer().write(segment)
-  val tag = buffer.readUtf8(2)
-  require(tag == "v1")
-  val offset = buffer.readInt()
-  val count = buffer.readInt()
-  return offset to count
-}
+/**
+ * It encodes the key range in utf8 for better cursor visibility in backfila.
+ * It's form:
+ * start = "v2:<segment_offset>/<segment_count>:<last_evaluated_key_json>"
+ * end = "v2:<segment_offset>/<segment_count>"
+ */
+@Singleton
+class DynamoDbKeyRangeCodec @Inject constructor(
+  @ForBackfila moshi: Moshi
+) {
+  private val adapter = moshi.adapter<Map<String, AttributeValue>>(
+      Types.newParameterizedType(Map::class.java, String::class.java, AttributeValue::class.java)
+  )
+
+  companion object {
+    const val VERSION = "v2"
+    const val INT_FORMAT = "%010d"
+  }
+
+  internal fun decodeKeyRange(keyRange: KeyRange): DynamoDbKeyRange {
+    val (startOffset, count, lastEvaluatedKey) = decodeSegment(keyRange.start)
+    val (endOffset, endCount, endLastEvaluatedKey) = decodeSegment(keyRange.end)
+
+    // start and end must have the same total segment count.
+    require(count == endCount)
 
-internal fun encodeKeyRange(
-  start: Int,
-  end: Int,
-  count: Int
-): KeyRange {
-  return KeyRange.Builder()
-      .start(encodeSegment(start, count))
-      .end(encodeSegment(end, count))
-      .build()
+    // end cannot have a lastEvaluatedKey
+    require(endLastEvaluatedKey == null)
+
+    return DynamoDbKeyRange(
+        startOffset,
+        endOffset,
+        count,
+        lastEvaluatedKey)
+  }
+
+  internal fun decodeSegment(segment: ByteString): Triple<Int, Int, Map<String, AttributeValue>?> {
```

Comment:
Shouldn’t this return a SegmentData?

Reply:
Yeah it should now yeah. It made less sense when I crammed in the batch size.